u_http.go

```diff
@@ -0,0 +1,565 @@
+// to cast the net.Conn to a *tls.Conn.
+//
+// This transport attempts to solve this issue by inspecting
+// the ALPN negotiated protocol and routing:
+//
+// - "h2" to a default constructed http2.Transport;
+//
+// - "http/1.1" to a default constructed http.Transport.
+//
+// Moreover, cleartext HTTP requests go to a default
+// constructed http.Transport.
+//
+// The zero initialized UHTTPTransport is valid and can be
+// used immediately. We will allocate internal variables when
+// we need them. As http.Transport, UHTTPTransport may have
+// idle connections, for which CloseIdleConnections can be used.
+//
+// You SHOULD NOT modify the public fields of this data
+// structure while it's being used, because that MAY
+// quite possibly lead to data races. Otherwise, it is
+// safe to call the methods of this struct from several
+// concurrent goroutines.
+type UHTTPTransport struct {
+	// TODO(bassosimone): what useful fields of an ordinary
+	// http.Transport should we also implement?
+
+	// DialContext is the optional dialer to dial connections
+	// just like the namesake field of http.Transport. If this
+	// dialer is set, we'll use it for dialing all conns.
+	DialContext func(ctx context.Context, network, address string) (net.Conn, error)
+
+	// Proxy is like the namesake field in http.Transport. If
+	// not initialized, or if it returns a nil URL, then there
+	// will be no proxying of connections. We support the
+	// same types of proxies as the stdlib for HTTP but we
+	// only support socks5 proxies for HTTPS/H2.
+	Proxy func(*http.Request) (*url.URL, error)
+
+	// TLSClientConfig contains optional UTLS configuration for
+	// this transport. We will default-construct a config
+	// instance if this field is not set. Otherwise,
+	// every dial attempt will use a Clone() of this field.
+	TLSClientConfig *Config
+
+	// TLSHandshakeTimeout is the optional maximum timeout we are
+	// willing to wait for the TLS handshake. If not set, we'll
+	// use a default TLS-handshake timeout of ten seconds.
+	TLSHandshakeTimeout time.Duration
+
+	// UTLSClientHelloID is the optional UTLS ClientHelloID
+	// that you would like to use with this transport. If
+	// nil, we will use utls.HelloFirefox_Auto.
+	UTLSClientHelloID *ClientHelloID
+
+	// cleartext is a transport for HTTP only. We will initialize
+	// this field on the first RoundTrip invocation.
+	cleartext uhttpCloseableTransport
+
+	// onlyDial is the transport used for dialing new
+	// connections. It will populate connCache and
+	// hostCache. We will initialize this field during
+	// the first invocation of RoundTrip.
+	onlyDial uhttpCloseableTransport
+
+	// https is a transport used only for the "http/1.1"
+	// ALPN. This transport does not perform any dial and
+	// only manages its cached persistent connections. We'll
+	// initialize it during the first RoundTrip call.
+	https uhttpCloseableTransport
+
+	// h2 is like https but for the "h2" ALPN.
+	h2 uhttpCloseableTransport
```

Comment:
nitpick: I'd prefer a more telling naming scheme for those transports, something like
`h1Cleartext`
`h1Secure`
`h2Secure`

Reply:
Sure!